edb/server/protocol/binary.pyx

```diff
@@ -1077,18 +1077,27 @@ cdef class EdgeConnection:
                                 config_ops = [
                                     config.Operation.from_json(r[0])
                                     for r in data]
-                        else:
+                        elif query_unit.output_format == FMT_NONE:
                             for sql in query_unit.sql:
                                 await conn.wait_for_command(
                                     ignore_data=True
                                 )
+                        else:
+                            for sql in query_unit.sql:
+                                await conn.wait_for_command(
+                                    ignore_data=False,
+                                    edgecon=self,
+                                )
 
                     if config_ops:
                         await _dbview.apply_config_ops(
                             conn,
                             config_ops)
 
-                    _dbview.on_success(query_unit, new_types)
+                    side_effects = _dbview.on_success(query_unit, new_types)
+                    if side_effects:
+                        # COMMIT MIGRATION breaks the atomicity as an exception
```

Comment:
Not sure this is correct.  Side effect signaling should only be done once the relevant transaction has been successfully committed.  `COMMIT MIGRATION` could be part of a transaction that can roll back, or, worse, have long-running statements which would mean that the recipients of side effect signals would not actually see the changes they've been signaled about.

Reply:
If `COMMIT MIGRATION` is part of a transaction, `dbview.on_success('COMMIT MIGRATION')` won't return any side effects, because it'll be a release savepoint command.

Oppositely, when there is no outer transaction, a script may have more than one `COMMIT MIGRATION` (with some other statements between migrations), that means we'll have 2 batches of side_effects. For example:

```
start migration to { module default { type A {} } };
populate migration;
commit migration;  -- side effects

create type C {};  -- start implicit transaction
start migration to { module default { type B {} } };  -- make implicit transaction "explicit"
populate migration;
commit migration;  -- side effects again
```

This is really breaking atomicity of scripts in different ways, so perhaps we should disallow migration commands outside of explicit transactions, regardless of script or not? Like, we should disallow doing this in Python:

```
>>> import edgedb
>>> c=edgedb.create_client('aa')
>>> c.execute('start migration to { module default { type A {} } }; populate migration;')
>>> c.execute('commit migration')
```

Reply:
Yes, we should disallow migration control outside of a transaction (implicit or explicit)